File: code/backend/app/evaluaciones/model/evaluacion.py
from datetime import datetime,timezone
from sqlalchemy import Column,Float,Integer,ForeignKey,DateTime,Text,UniqueConstraint,CheckConstraint,Boolean
from sqlalchemy.orm import relationship
from database import Base

class Evaluacion(Base):
    __tablename__ = "evaluacion"
    id = Column(Integer, primary_key = True)
    nro_evaluacion = Column(Integer, nullable=False)
    altura = Column(Integer, nullable=False)
    peso = Column(Float, nullable=False)
    masa_muscular = Column(Float, nullable=False)
    observacion = Column(Text,nullable=True)
    fecha_registro = Column(DateTime,nullable=False, default=lambda: datetime.now(timezone.utc))
    esta_activo = Column(Boolean, nullable=False,default=True)
    paciente_id = Column(Integer, ForeignKey("paciente.id"),nullable=False)

    __table_args__ = (
        UniqueConstraint("paciente_id", "nro_evaluacion", name="unico_paciente_nro_evaluacion"),
        CheckConstraint("masa_muscular >= 0 AND masa_muscular <=100", name="chequeo_masa_muscular_rango"),
        CheckConstraint("peso > 0", name="chequeo_peso_positivo"),
        CheckConstraint("altura > 0", name="chequeo_altura_positiva"),
    )

    # relacion con paciente
    paciente = relationship("Paciente",back_populates="evaluaciones")    

File: code/frontend_flutter/lib/models/evaluacion.dart
class Evaluacion {
  final int id;
  final int nro_evaluacion;
  final int altura;
  final double peso;
  final double imc;
  final double masa_muscular;
  final String? observacion;
  final DateTime fechaRegistro;
  final int pacienteId;
  final bool esta_activo;

  Evaluacion({
    required this.id,
    required this.nro_evaluacion,
    required this.altura,
    required this.peso,
    required this.imc,
    required this.masa_muscular,
    required this.observacion,
    required this.fechaRegistro,
    required this.pacienteId,
    required this.esta_activo,
  });

  factory Evaluacion.fromJson(Map<String, dynamic> json) {
    return Evaluacion(
      id: json['id'] as int,
      nro_evaluacion: json['nro_evaluacion'] as int,
      altura: json['altura'] as int,
      peso: json['peso'] as double,
      imc: (json['imc'] as num).toDouble(),
      masa_muscular: json['masa_muscular'] as double,
      observacion: json['observacion'] as String?,
      fechaRegistro: DateTime.parse(json['fecha_registro'] as String),
      pacienteId: json['paciente_id'] as int,
      esta_activo: json['esta_activo'] as bool? ?? true,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'nro_evaluacion': nro_evaluacion,
      'altura': altura,
      'peso': peso,
      'imc': imc,
      'masa_muscular': masa_muscular,
      'observacion': observacion,
      'fecha_registro': fechaRegistro.toIso8601String(),
      'paciente_id': pacienteId,
      'esta_activo': esta_activo,
    };
  }
}
